qiskit_experiments/framework/experiment_data.py

```diff
@@ -985,27 +988,58 @@ def _add_result_data(self, result: Result, job_id: Optional[str] = None) -> None
-                data["shots"] = expr_result.shots
-                data["meas_level"] = expr_result.meas_level
-                if hasattr(expr_result, "meas_return"):
-                    data["meas_return"] = expr_result.meas_return
-                self._result_data.append(data)
+        if hasattr(result, "results"):
+            # backend run results
+            if job_id is None:
+                job_id = result.job_id
+            if job_id not in self._jobs:
+                self._jobs[job_id] = None
+                self.job_ids.append(job_id)
+            with self._result_data.lock:
+                # Lock data while adding all result data
+                for i, _ in enumerate(result.results):
+                    data = result.data(i)
+                    data["job_id"] = job_id
+                    if "counts" in data:
+                        # Format to Counts object rather than hex dict
+                        data["counts"] = result.get_counts(i)
+                    expr_result = result.results[i]
+                    if hasattr(expr_result, "header") and hasattr(expr_result.header, "metadata"):
+                        data["metadata"] = expr_result.header.metadata
+                    data["shots"] = expr_result.shots
+                    data["meas_level"] = expr_result.meas_level
+                    if hasattr(expr_result, "meas_return"):
+                        data["meas_return"] = expr_result.meas_return
+                    self._result_data.append(data)
+        else:
+            # sampler results
+            if job_id is None:
+                raise QiskitError("job_id must be provided, not available in the sampler result")
+            if job_id not in self._jobs:
+                self._jobs[job_id] = None
+                self.job_ids.append(job_id)
+            with self._result_data.lock:
+                # Lock data while adding all result data
+                # Sampler results are a list
+                for i, _ in enumerate(result):
+                    data = {}
+                    # convert to a Sampler Pub Result (can remove this later when the bug is fixed)
+                    testres = SamplerPubResult(result[i].data, result[i].metadata)
+                    data["job_id"] = job_id
+                    if isinstance(testres.data[next(iter(testres.data))], BitArray):
+                        # bit results so has counts
+                        data["meas_level"] = 2
+                        data["meas_return"] = "avg"
+                        # join the data
+                        data["counts"] = testres.join_data(testres.data.keys()).get_counts()
+                        # number of shots
+                        data["shots"] = testres.data[next(iter(testres.data))].num_shots
+                    else:
+                        raise QiskitError("Sampler with meas level 1 support TBD")
+
+                    data["metadata"] = testres.metadata["circuit_metadata"]
```

Comment:
```suggestion
                    if hasattr(testres.metadata, "circuit_metadata"):
                        data["metadata"] = testres.metadata["circuit_metadata"]
```

Reply:
Actually, this appears insufficient to get metadata to flow through. The only alternative I have come up with so far is to pass the `Job` itself rather than the job id to this method, and then grab the metadata from the pubs provided as job inputs. i.e. something like:
```python
if hasattr(testres.metadata, "circuit_metadata"):
    data["metadata"] = testres.metadata["circuit_metadata"]
else:
    first_pub = job.inputs["pubs"][0]
    first_circuit = first_pub[0]
    data["metadata"] = first_circuit.metadata
```

Reply:
It turns out this solution ^^ doesn't work either. I will continue investigating.